drivers/gpio/gpio_lpc11u6x.c

```diff
@@ -0,0 +1,655 @@
+ * shared between all the ports (IRQ lines, clock).
+ */
+
+static const struct gpio_lpc11u6x_shared gpio_lpc11u6x_shared = {
+	.clock_controller_name = DT_LABEL(DT_INST_PHANDLE(0, clocks)),
+	.clock_subsys = (clock_control_subsys_t) DT_INST_PHA(0, clocks, clkid),
+	.gpio_base = DT_INST_REG_ADDR_BY_IDX(0, 0),
+	.syscon_base = DT_INST_REG_ADDR_BY_IDX(0, 1),
+	.nirqs = DT_NUM_IRQS(DT_DRV_INST(0)),
+};
+
+#define IRQ_INIT(n)							\
+do {							                \
+	IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, n, irq),			\
+		    DT_INST_IRQ_BY_IDX(0, n, priority),			\
+		    gpio_lpc11u6x_isr, &gpio_lpc11u6x_shared, 0);	\
+	irq_enable(DT_INST_IRQ_BY_IDX(0, n, irq));			\
+} while (0)
+
+static int gpio_lpc11u6x_init(struct device *dev)
+{
+	const struct gpio_lpc11u6x_config *config = DEV_CFG(dev);
+	struct gpio_lpc11u6x_data *data = DEV_DATA(dev);
+	struct device *clock_dev;
+	int ret;
+	static bool gpio_ready;
+
+	/* Retrieve pinmux device. */
+	data->pinmux_dev = device_get_binding(config->pinmux_name);
+	if (!data->pinmux_dev) {
+		return -EINVAL;
+	}
+
+	/* Initialize shared resources only once. */
+	if (gpio_ready) {
+		return 0;
+	}
+
+	/* Enable GPIO and PINT clocks. */
+	clock_dev = device_get_binding(config->shared->clock_controller_name);
+	if (!clock_dev) {
+		return -ENODEV;
+	}
+
+	ret = clock_control_on(clock_dev, config->shared->clock_subsys);
+	if (ret < 0) {
+		return ret;
+	}
+
+#if DT_INST_IRQ_HAS_IDX(0, 0)
+	IRQ_INIT(0);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 1)
+	IRQ_INIT(1);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 2)
+	IRQ_INIT(2);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 3)
+	IRQ_INIT(3);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 4)
+	IRQ_INIT(4);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+	IRQ_INIT(5);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+	IRQ_INIT(6);
+#endif
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+	IRQ_INIT(7);
+#endif
+	gpio_ready = true;
+
+	return 0;
+}
+
+#define GPIO_LPC11U6X_INIT(id)						\
+static const struct gpio_lpc11u6x_config				\
+			gpio_lpc11u6x_config_##id = {			\
+	.common = {							\
+		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_NODE(	\
+					DT_NODELABEL(gpio##id)),	\
+	},								\
+	.shared = &gpio_lpc11u6x_shared,				\
+	.port_num = id,							\
+	.pinmux_name = DT_LABEL(DT_PHANDLE(DT_NODELABEL(gpio##id),	\
+				pinmux_port)),				\
+	.ngpios = DT_PROP(DT_NODELABEL(gpio##id), ngpios),		\
+};									\
+									\
+static struct gpio_lpc11u6x_data gpio_lpc11u6x_data_##id;		\
+									\
+DEVICE_AND_API_INIT(gpio_lpc11u6x_##id,					\
+		    DT_LABEL(DT_NODELABEL(gpio##id)),			\
+		    &gpio_lpc11u6x_init,				\
+		    &gpio_lpc11u6x_data_##id,				\
+		    &gpio_lpc11u6x_config_##id,				\
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		    &gpio_lpc11u6x_driver_api)
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio0), okay)
+GPIO_LPC11U6X_INIT(0);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio1), okay)
+GPIO_LPC11U6X_INIT(1);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio2), okay)
+GPIO_LPC11U6X_INIT(2);
+#endif
```

Comment:
```suggestion
DT_INST_FOREACH_STATUS_OKAY(GPIO_LPC11U6X_INIT)
```

Reply:
I spent some time this morning trying to do that. The result is I'd rather prefer not :) The way the hardware works, any interruption can be mapped on any GPIO port. So when an interrupt happens, the handler has to retrieve first the GPIO port number and next the "struct device" associated with. To do that, having the GPIO device X associated with the GPIO port X helps a lot. With DT_INST_FOREACH_STATUS_OKAY we can't assume that. If for example the GPIO port 0 node is disabled in DT (let's say by an application overlay), then instance 0 will be GPIO port 1.